### PR TITLE
[Inductor][MPS] Fix half-precision type mismatches in Metal shader codegen (#176436)

### PR DESCRIPTION
**Pull request by malfet**
Metal Shading Language rejects implicit float-to-bfloat conversions, so
bare float literals like `0.0` in generated shaders cause compilation
failures when the target variable is `bfloat` (or `half`). Three codegen
methods were affected:

- `constant()` ignored its `dtype` parameter and returned raw literals.
- `masked()` assigned a bare literal in the else-branch (`} else tmp = 0.0;`).
- `where()` passed a bare literal through the ternary without casting.

All three now emit `static_cast<bfloat>(...)` / `static_cast<half>(...)`
where needed. Tests added for half-precision constants, reductions, and
conditionals.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/176436
Approved by: https://github.com/malfet

Test plan: Run `python -c "import torch;F=torch.nn.functional;print(torch.compile(lambda x: F.pad(F.gelu(x), [1, 0]))(torch.randn(4, device='mps', dtype=torch.bfloat16)))"`

(cherry picked from commit 3b161e7a756798e6eb1ab096f4ef1232d163a68d)

### CODE REVIEW

**Review by atalman — APPROVED** [review]
lgtm

**Comment by yangw-dev** [comment]
# Verify the bug is fixed in release 2.11.0.

## Repro Summary: Metal Shading Language bfloat/half implicit float conversion

**Test:** Ran the MRE from the bug report using `torch.compile` with `bfloat16` on MPS.

**Result: Fixed**
- Shader compiles successfully without implicit float-to-bfloat conversion errors
- No Metal shader compilation failures
- Test wheel location: https://dev-discuss.pytorch.org/t/pytorch-2-11-final-rc-available/3326

if not resolved:
Metal shader compilation would fail with an implicit conversion error:
```
error: cannot implicitly convert type 'float' to 'bfloat'
```

## Local Reproduce

### Setup
```bash
uv venv mps-repro
source mps-repro/bin/activate

uv pip install torch==2.11.0 torchvision torchaudio --index-url https://download.pytorch.org/whl/test/cpu

python -c "import torch; print(torch.__version__)"
# output: 2.11.0

python -c "import torch;F=torch.nn.functional;print(torch.compile(lambda x: F.pad(F.gelu(x), [1, 0]))(torch.randn(4, device='mps', dtype=torch.bfloat16)))"
```

### Output
```
tensor([ 0.0000,  0.3047,  0.2344, -0.1680,  1.1797], device='mps:0',
       dtype=torch.bfloat16)
```

---

## Versions

```
PyTorch version: 2.11.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.3.1 (arm64)
GCC version: Could not collect
Clang version: 17.0.0 (clang-1700.6.4.2)
CMake version: version 4.3.0
Libc version: N/A

Python version: 3.13.12 (main, Feb 3 2026, 17:53:27) [Clang 17.0.0 (clang-1700.6.3.2)] (64-bit runtime)
Python platform: macOS-26.3.1-arm64-arm-64bit-Mach-O
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU: Apple M1 Pro

Versions of relevant libraries:
[pip3] Could not collect
[conda] Could not collect
```
